packages/@lwc/integration-karma/test/rendering/callback-invocation-order/index.spec.js

```diff
@@ -297,3 +299,90 @@ it('should invoke callbacks on the right order when multiple templates are used
             ]);
         });
 });
+
+if (process.env.NATIVE_SHADOW) {
+    // #TODO[3827]: These regression tests can be removed once the bug is resolved.
+    it('regression test (#3827) - shadow DOM should maintain callback invocation order for duplicate slots across conditional branches when rerender triggered on conditional slot component', () => {
+        const container = createElement('x-shadow-container-multiple-conditionals', {
+            is: ShadowContainerMultipleConditionals,
+        });
+        document.body.appendChild(container);
+
+        const currentLeafName = container.getLeaf().name;
+        expect(window.timingBuffer).toEqual([
+            'shadowContainer:connectedCallback',
+            'shadowSlot:connectedCallback',
+            `leaf:${currentLeafName}:connectedCallback`,
+        ]);
+
+        resetTimingBuffer();
+
+        const slotCmp = container.getSlotCmp();
+        slotCmp.showIf = true;
+
+        return Promise.resolve()
+            .then(() => {
+                expect(window.timingBuffer).toEqual([]);
+
+                slotCmp.showIf = false;
+                slotCmp.showElseIf = true;
+            })
+            .then(() => {
+                expect(window.timingBuffer).toEqual([]);
+                document.body.removeChild(container);
+            });
+    });
+}
+
+it('regression test (#3827) - light DOM should maintain callback invocation order for duplicate slots across conditional branches when rerender triggered on conditional slot component', () => {
+    // #TODO[3827]: These regression tests can be removed once the bug is resolved.
+    const container = createElement('x-light-container-multiple-conditionals', {
+        is: LightContainerMultipleConditionals,
+    });
+    document.body.appendChild(container);
+
+    let currentLeafName = container.getLeaf().name;
+    expect(window.timingBuffer).toEqual([
+        'lightContainer:connectedCallback',
+        'lightSlot:connectedCallback',
+        `leaf:${currentLeafName}:connectedCallback`,
+    ]);
```

Comment:
These tests are a little strange, but I created them to capture the current behavior and ensure they match what happens in 3.x.x.

Reply:
Note that this bug only occurs if the conditional rendering occurs on the same component as the `<slot>` element.

The bug also only occurs when the condition is only triggered through the component containing the `<slot>`.

If the condition depends on a value being passed from a parent component, the slots will render correctly.